File: Lab/SEM4/MPP/lab9/moto_c/src/Client/sync/SocketNotifier.cs
using moto_c.Common.sync;

namespace moto_c.sync;

using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Sockets;
using System.Threading;

public class SocketNotifier
{
    private readonly int port;
    private readonly string serverHost;
    private readonly int serverPort;

    private readonly List<Action<string>> listeners = new();
    private StreamWriter? writer;
    private bool started = false;

    public SocketNotifier(int port, string serverHost, int serverPort)
    {
        this.port = port;
        this.serverHost = serverHost;
        this.serverPort = serverPort;
    }

    public void OnUpdate(Action<string> callback)
    {
        listeners.Add(callback);
    }

    public void Start()
    {
        if (started) return;
        started = true;

        new Thread(() =>
        {
            try
            {
                var socket = new TcpClient(serverHost, serverPort);
                var stream = socket.GetStream();

                writer = new StreamWriter(stream) { AutoFlush = true };

                Console.WriteLine($"[Port {port}] Connected to SyncServer at {serverHost}:{serverPort}");

                var reader = new StreamReader(stream);

                string? msg;
                while ((msg = reader.ReadLine()) != null)
                {
                    foreach (var listener in listeners)
                    {
                        listener(msg);
                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
            }
        })
        {
            IsBackground = true
        }.Start();
    }

    public void NotifyPeer(string message)
    {
        if (writer != null)
        {
            Console.WriteLine($"[Port {port}] sending: {message}");
            writer.WriteLine(message);
        }
        else
        {
            Console.Error.WriteLine($"[Port {port}] Not connected to server yet");
        }
    }

    public int GetPort()
    {
        return port;
    }

    public void Respond(string requestType, ResponseType type, string message)
    {
        if (writer != null)
        {
            var response = new Response(type, requestType, message);
            writer.WriteLine(response.ToString());
        }
        else
        {
            Console.Error.WriteLine($"[Port {port}] Not connected to server yet, cannot respond");
        }
    }
}

File: Lab/SEM4/MPP/lab7/moto_java/client/src/main/java/org/zdroba/sync/SocketNotifier.java
package org.zdroba.sync;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;

public class SocketNotifier {
    private final int port;
    private final String serverHost;
    private final int serverPort;
    private final List<Consumer<String>> listeners = new CopyOnWriteArrayList<>();
    private PrintWriter out;
    private boolean started = false;

    public SocketNotifier(int port, String serverHost, int serverPort) {
        this.port = port;
        this.serverHost = serverHost;
        this.serverPort = serverPort;
    }

    public void onUpdate(Consumer<String> callback) {
        listeners.add(callback);
    }

    public void start() {
        if (started) return;
        started = true;

        new Thread(() -> {
            try {
                Socket socket = new Socket(serverHost, serverPort);
                out = new PrintWriter(socket.getOutputStream(), true);
                System.out.println("[Port " + port + "] Connected to SyncServer at " + serverHost + ":" + serverPort);

                BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
                String msg;
                while ((msg = in.readLine()) != null) {
                    final String received = msg;
                    for (Consumer<String> listener : listeners) {
                        listener.accept(received);
                    }
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }, "SocketNotifier-" + port).start();
    }

    public void notifyPeer(String message) {
        if (out != null) {
            System.out.println("[Port " + port + "] sending: " + message);
            out.println(message);
        } else {
            System.err.println("[Port " + port + "] Not connected to server yet");
        }
    }

    public int getPort() {
        return port;
    }

    public void respond(String requestType, ResponseType type, String message) {
        if(out != null) {
            Response response = new Response(type, requestType, message);
            out.println(response.toString());
        } else {
            System.err.println("[Port " + port + "] Not connected to server yet, cannot respond");
        }
    }
}